docfiles/pxtweb/cookieCompliance.ts

```diff
@@ -0,0 +1,217 @@
+    }
+
+    interface HttpResponse {
+        status: number;
+        body: string;
+    }
+
+    interface Callback<T> {
+        (err?: any, res?: T): void;
+    }
+
+    export function initAnalyticsAsync() {
+        getCookieBannerAsync(detectLocale(), document.domain, (bannerErr, info) => {
+            if (bannerErr || info.Error) {
+                // Start app insights, just don't drop any cookies
+                initializeAppInsights(false);
+                return;
+            }
+
+            // Clear the cookies if the consent is too old, mscc won't do it automatically
+            if (isConsentExpired(info.CookieName, info.MinimumConsentDate)) {
+                const definitelyThePast = new Date(0).toUTCString();
+                document.cookie = `ai_user=; expires=${definitelyThePast}`;
+                document.cookie = `ai_session=; expires=${definitelyThePast}`;
+                document.cookie = `${info.CookieName}=0; expires=${definitelyThePast}`;
+            }
+
+            let bannerDiv = document.getElementById("cookiebanner");
+            if (!bannerDiv) {
+                bannerDiv = document.createElement("div");
+                document.body.insertBefore(bannerDiv, document.body.firstChild);
+            }
+
+            // The markup is trusted because it's from our backend, so it shouldn't need to be scrubbed
+            bannerDiv.innerHTML = info.Markup;
+
+            if (info.Css && info.Css.length) {
+                info.Css.forEach(injectStylesheet)
+            }
+
+            all(info.Js || [], injectScriptAsync, msccError => {
+                initializeAppInsights(!msccError && typeof mscc !== "undefined" && mscc.hasConsent());
+            });
+        });
+    }
+
+    function detectLocale() {
+        // Intentionally ignoring the default locale in the target settings and the language cookie
+        // Warning: app.tsx overwrites the hash after reading the language so this needs
+        // to be called before that happens
+        const mlang = /(live)?lang=([a-z]{2,}(-[A-Z]+)?)/i.exec(window.location.href);
+        return mlang ? mlang[2] : ((navigator as any).userLanguage || navigator.language);
+    }
+
+    function getCookieBannerAsync(domain: string, locale: string, cb: Callback<CookieBannerInfo>) {
+        httpGetAsync(`https://makecode.com/api/mscc/${domain}/${locale}`, function(err, resp) {
+            if (err) {
+                cb(err);
+                return;
+            }
+
+            if (resp.status === 200) {
+                try {
+                    const info = JSON.parse(resp.body);
+                    cb(undefined, info as CookieBannerInfo);
+                    return;
+                }
+                catch (e) {
+                    cb(new Error("Bad response from server: " + resp.body))
+                    return;
+                }
+            }
+            cb(new Error("didn't get 200 response: " + resp.status + " " + resp.body));
```

Comment:
Handle failure gracefully.

Reply:
What's un-graceful about this? I'm passing an error, not throwing it.